apmproxy/apmserver.go

```diff
@@ -74,15 +76,37 @@ func (c *Client) FlushAPMData(ctx context.Context) {
 		return
 	}
 	c.logger.Debug("Flush started - Checking for agent data")
+
+	// Flush agent data first to make sure metadata is available if possible
+	for i := len(c.AgentDataChannel); i > 0; i-- {
+		data := <-c.AgentDataChannel
+		if err := c.forwardAgentData(ctx, data); err != nil {
+			c.logger.Errorf("Error sending to APM Server, skipping: %v", err)
+		}
+	}
+
+	// If metadata still not available then fail fast
+	if c.batch == nil {
+		c.logger.Warnf("Metadata not available at flush, skipping sending lambda data to APM Server")
+		return
+	}
+
+	// Flush lambda data
 	for {
 		select {
-		case agentData := <-c.DataChannel:
-			c.logger.Debug("Flush in progress - Processing agent data")
-			if err := c.PostToApmServer(ctx, agentData); err != nil {
+		case <-ctx.Done():
+			c.logger.Debugf("Failed to flush completely, may result in data drop")
```

Comment:
```suggestion
			c.logger.Debugf("Failed to flush completely, may result in data drop")
			return
```

IIANM we must return here or we'll enter an infinite loop, as we'll only hit the default case if none of the other cases can be run.

Reply:
OMG, thanks for catching this.

Reply:
Fixed.